drivers/hv/mshv_vtl_main.c

```diff
@@ -895,12 +894,17 @@ noinline void mshv_vtl_return_tdx(void)
 	rdmsrl(MSR_SYSCALL_MASK, tdx_vp_state->msr_sfmask);
 	rdmsrl(MSR_TSC_AUX, tdx_vp_state->msr_tsc_aux);
 
-	/* Restore VTL2's syscall registers & MSRs */
-	wrmsrl(MSR_KERNEL_GS_BASE, per_cpu->l1_msr_kernel_gs_base);
-	wrmsrl(MSR_STAR, per_cpu->l1_msr_star);
-	wrmsrl(MSR_LSTAR, per_cpu->l1_msr_lstar);
-	wrmsrl(MSR_SYSCALL_MASK, per_cpu->l1_msr_sfmask);
-	wrmsrl(MSR_TSC_AUX, per_cpu->l1_msr_tsc_aux);
+	per_cpu->l1_msr_kernel_gs_base_curr = tdx_vp_state->msr_kernel_gs_base;
+	per_cpu->l1_msr_star_curr = tdx_vp_state->msr_star;
+	per_cpu->l1_msr_lstar_curr = tdx_vp_state->msr_lstar;
+	per_cpu->l1_msr_sfmask_curr = tdx_vp_state->msr_sfmask;
+	per_cpu->l1_msr_tsc_aux_curr = tdx_vp_state->msr_tsc_aux;
+
+	if (!per_cpu->urn_registered) {
+		per_cpu->mshv_urn.on_user_return = mshv_vtl_on_user_return;
+		user_return_notifier_register(&per_cpu->mshv_urn);
```

Comment:
i assume this callback is per-cpu?

Reply:
The comment around the implementation suggests it is per-cpu. As MSR are per-cpu, the call needs to be on that cpu.
*
 * Request a notification when the current cpu returns to userspace.  Must be
 * called in atomic context.  The notifier will also be called in atomic
 * context.
 */